contrib/mesos/pkg/scheduler/podtask/procurement.go

```diff
@@ -50,77 +72,203 @@ type AllOrNothingProcurement []Procurement
 
 // Procure runs each Procurement in the receiver list. The first Procurement func that
 // fails triggers T.Reset() and the error is returned, otherwise returns nil.
-func (a AllOrNothingProcurement) Procure(t *T, offer *mesos.Offer) error {
+func (a AllOrNothingProcurement) Procure(t *T, offer *mesos.Offer, n *api.Node, spec *Spec) error {
 	for _, p := range a {
-		if err := p(t, offer); err != nil {
-			t.Reset()
+		err := p.Procure(t, offer, n, spec)
+		if err != nil {
 			return err
 		}
 	}
 	return nil
 }
 
-// ValidateProcurement checks that the offered resources are kosher, and if not panics.
-// If things check out ok, t.Spec is cleared and nil is returned.
-func ValidateProcurement(t *T, offer *mesos.Offer) error {
-	if offer == nil {
-		//programming error
-		panic("offer details are nil")
-	}
-	t.Spec = Spec{}
-	return nil
+// NewNodeProcurement returns a Procurement that checks whether the given pod task and offer
+// have valid node informations available and wehther the pod spec node selector matches
+// the pod labels.
+// If the check is successfull the slave ID and assigned slave is set in the given Spec.
+func NewNodeProcurement() Procurement {
+	return ProcurementFunc(func(t *T, offer *mesos.Offer, n *api.Node, spec *Spec) error {
+		// if the user has specified a target host, make sure this offer is for that host
+		if t.Pod.Spec.NodeName != "" && offer.GetHostname() != t.Pod.Spec.NodeName {
+			return fmt.Errorf(
+				"NodeName %q does not match offer hostname %q",
+				t.Pod.Spec.NodeName, offer.GetHostname(),
+			)
+		}
+
+		// check the NodeSelector
+		if len(t.Pod.Spec.NodeSelector) > 0 {
+			if n.Labels == nil {
+				return fmt.Errorf(
+					"NodeSelector %v does not match empty labels of pod %s/%s",
+					t.Pod.Spec.NodeSelector, t.Pod.Namespace, t.Pod.Name,
+				)
+			}
+			selector := labels.SelectorFromSet(t.Pod.Spec.NodeSelector)
+			if !selector.Matches(labels.Set(n.Labels)) {
+				return fmt.Errorf(
+					"NodeSelector %v does not match labels %v of pod %s/%s",
+					t.Pod.Spec.NodeSelector, t.Pod.Labels, t.Pod.Namespace, t.Pod.Name,
+				)
+			}
+		}
+
+		spec.SlaveID = offer.GetSlaveId().GetValue()
+		spec.AssignedSlave = offer.GetHostname()
+
+		return nil
+	})
 }
 
-// NodeProcurement updates t.Spec in preparation for the task to be launched on the
-// slave associated with the offer.
-func NodeProcurement(t *T, offer *mesos.Offer) error {
-	t.Spec.SlaveID = offer.GetSlaveId().GetValue()
-	t.Spec.AssignedSlave = offer.GetHostname()
-	return nil
+// NewPodResourcesProcurement converts k8s pod cpu and memory resource requirements into
+// mesos resource allocations.
+func NewPodResourcesProcurement() Procurement {
+	return ProcurementFunc(func(t *T, offer *mesos.Offer, _ *api.Node, spec *Spec) error {
+		// TODO(sttts): fall back to requested resources if resource limit cannot be fulfilled by the offer
```

Comment:
In fact, I was just thinking about accepting offers with resources between request and limit, as a fallback. But of course something dynamic would be even better.